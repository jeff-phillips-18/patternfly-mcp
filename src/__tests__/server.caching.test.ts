import { memo } from '../server.caching';

describe('memo', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it.each([
    {
      description: 'sync',
      options: { cacheLimit: 2 },
      params: [[], [], [1], [1], [2], [2], [3], [3], [1]]
    },
    {
      description: 'sync errors',
      options: { cacheLimit: 2, cacheErrors: true },
      params: [[, true], [, true], [4, true], [4, true], [5, true], [5, true], [6, true], [6, true], [4, true]]
    },
    {
      description: 'sync errors NOT cached',
      options: { cacheLimit: 2, cacheErrors: false },
      params: [[7, true], [7, true], [8], [8], [9, true], [9, true], [7, true]]
    },
    {
      description: 'bypass memoization when cacheLimit is zero',
      options: { cacheLimit: 0 },
      params: [[], [], [1], [1], [2, true], [2, true]]
    }
  ])('should memoize a function, $description', ({ options, params }) => {
    const log: { type: string; value: () => string; cache: string[] }[] = [];
    const debug = (response: any) => log.push(response);

    const memoized = memo(
      (str, isError = false) => {
        const arr = ['lorem', 'ipsum', 'dolor', 'sit'];
        const randomStr = Math.floor(Math.random() * arr.length);
        const genStr = `${arr[randomStr]}-${str || '[EMPTY]'}`;

        if (isError) {
          throw new Error(genStr);
        }

        return genStr;
      },
      { debug, ...options }
    );

    for (const param of params) {
      try {
        memoized(...param as [unknown, unknown]);
      } catch {}
    }

    const updatedLog = [];

    for (const { type, value, cache } of log) {
      let successValue;
      let errorValue;

      try {
        successValue = value();
      } catch (e) {
        const error = e as Error;
        errorValue = error.message;
      }

      successValue = successValue?.split?.('-')[1];
      errorValue = errorValue?.split?.('-')[1];

      updatedLog.push({ type, successValue, errorValue, cacheLength: cache.length });
    }

    expect(updatedLog).toMatchSnapshot();
  });

  it.each([
    {
      description: 'async',
      options: { cacheLimit: 2 },
      params: [[], [], [1], [1], [2], [2], [3], [3], [1]]
    },
    {
      description: 'async errors',
      options: { cacheLimit: 2, cacheErrors: true },
      params: [[, true], [, true], [4, true], [4, true], [5, true], [5, true], [6, true], [6, true], [4, true]]
    },
    {
      description: 'async errors NOT cached',
      options: { cacheLimit: 2, cacheErrors: false },
      params: [[7, true], [7, true], [8], [8], [9, true], [9, true], [7, true]]
    },
    {
      description: 'async bypass memoization when cacheLimit is zero',
      options: { cacheLimit: 0 },
      params: [[], [], [1], [1], [2, true], [2, true]]
    }
  ])('should memoize a function, $description', async ({ options, params }) => {
    const log: { type: string; value: () => Promise<string>; cache: string[] }[] = [];
    const debug = (response: any) => log.push(response);
    const memoized = memo(
      async (str, isError = false) => {
        const arr = ['lorem', 'ipsum', 'dolor', 'sit'];
        const randomStr = Math.floor(Math.random() * arr.length);
        const genStr = `${arr[randomStr]}-${str || '[EMPTY]'}`;

        if (isError) {
          throw new Error(genStr);
        }

        return genStr;
      },
      { debug, ...options }
    );

    try {
      await Promise.all(params.map(param => memoized(...param as [unknown, unknown])));
    } catch {}

    const updatedLog = [];

    for (const { type, value, cache } of log) {
      let successValue;
      let errorValue;

      try {
        successValue = await value();
      } catch (e) {
        const error = e as Error;
        errorValue = error.message;
      }

      successValue = successValue?.split?.('-')[1];
      errorValue = errorValue?.split?.('-')[1];

      updatedLog.push({ type, successValue, errorValue, cacheLength: cache.length });
    }

    expect(updatedLog).toMatchSnapshot();
  });

  it.each([
    {
      description: 'async',
      options: { cacheLimit: 3, expire: 10 },
      paramsOne: [[], [], [1], [1], [2], [2]],
      paramsTwo: [[3, true], [3, true], [4, true], [4, true], [5, true], [5, true]],
      pause: 70
    }
  ])('should clear cache on inactivity, $description', async ({ options, paramsOne, paramsTwo, pause }) => {
    const log: { type: string; value: () => Promise<string>; cache: string[] }[] = [];
    const debug = (response: any) => log.push(response);
    const memoized = memo(async (str, isError = false) => {
      const arr = ['lorem', 'ipsum', 'dolor', 'sit'];
      const randomStr = Math.floor(Math.random() * arr.length);
      const genStr = `${arr[randomStr]}-${str || '[EMPTY]'}`;

      if (isError) {
        throw new Error(genStr);
      }

      return genStr;
    }, { debug, ...options });

    try {
      await Promise.all(paramsOne.map(param => memoized(...param as [unknown])));
    } catch {}

    jest.advanceTimersByTime(pause);

    try {
      await Promise.all(paramsTwo.map(param => memoized(...param as [unknown, unknown])));
    } catch {}

    jest.advanceTimersByTime(pause);

    const updatedLog = [];

    for (const { cache } of log) {
      updatedLog.push(cache.filter(Boolean).length);
    }

    expect(updatedLog).toMatchSnapshot('cache list length');
  });
});
