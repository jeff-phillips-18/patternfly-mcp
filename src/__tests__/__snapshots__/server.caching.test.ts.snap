// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`memo should clear cache on inactivity, async: cache list length 1`] = `
[
  2,
  2,
  4,
  4,
  6,
  6,
  2,
  2,
  4,
  4,
  6,
  6,
]
`;

exports[`memo should memoize a function, async 1`] = `
[
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "[EMPTY]",
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "[EMPTY]",
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "1",
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "1",
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "2",
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "2",
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "3",
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "3",
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "1",
    "type": "memo promise",
  },
]
`;

exports[`memo should memoize a function, async bypass memoization when cacheLimit is zero 1`] = `
[
  {
    "cacheLength": 0,
    "errorValue": undefined,
    "successValue": "[EMPTY]",
    "type": "memo bypass",
  },
  {
    "cacheLength": 0,
    "errorValue": undefined,
    "successValue": "[EMPTY]",
    "type": "memo bypass",
  },
  {
    "cacheLength": 0,
    "errorValue": undefined,
    "successValue": "1",
    "type": "memo bypass",
  },
  {
    "cacheLength": 0,
    "errorValue": undefined,
    "successValue": "1",
    "type": "memo bypass",
  },
  {
    "cacheLength": 0,
    "errorValue": "2",
    "successValue": undefined,
    "type": "memo bypass",
  },
  {
    "cacheLength": 0,
    "errorValue": "2",
    "successValue": undefined,
    "type": "memo bypass",
  },
]
`;

exports[`memo should memoize a function, async errors 1`] = `
[
  {
    "cacheLength": 4,
    "errorValue": "[EMPTY]",
    "successValue": undefined,
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": "[EMPTY]",
    "successValue": undefined,
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": "4",
    "successValue": undefined,
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": "4",
    "successValue": undefined,
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": "5",
    "successValue": undefined,
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": "5",
    "successValue": undefined,
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": "6",
    "successValue": undefined,
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": "6",
    "successValue": undefined,
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": "4",
    "successValue": undefined,
    "type": "memo promise",
  },
]
`;

exports[`memo should memoize a function, async errors NOT cached 1`] = `
[
  {
    "cacheLength": 4,
    "errorValue": "7",
    "successValue": undefined,
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": "7",
    "successValue": undefined,
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "8",
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "8",
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": "9",
    "successValue": undefined,
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": "9",
    "successValue": undefined,
    "type": "memo promise",
  },
  {
    "cacheLength": 4,
    "errorValue": "7",
    "successValue": undefined,
    "type": "memo promise",
  },
]
`;

exports[`memo should memoize a function, bypass memoization when cacheLimit is zero 1`] = `
[
  {
    "cacheLength": 0,
    "errorValue": undefined,
    "successValue": "[EMPTY]",
    "type": "memo bypass",
  },
  {
    "cacheLength": 0,
    "errorValue": undefined,
    "successValue": "[EMPTY]",
    "type": "memo bypass",
  },
  {
    "cacheLength": 0,
    "errorValue": undefined,
    "successValue": "1",
    "type": "memo bypass",
  },
  {
    "cacheLength": 0,
    "errorValue": undefined,
    "successValue": "1",
    "type": "memo bypass",
  },
]
`;

exports[`memo should memoize a function, sync 1`] = `
[
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "[EMPTY]",
    "type": "memo",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "[EMPTY]",
    "type": "memo",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "1",
    "type": "memo",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "1",
    "type": "memo",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "2",
    "type": "memo",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "2",
    "type": "memo",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "3",
    "type": "memo",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "3",
    "type": "memo",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "1",
    "type": "memo",
  },
]
`;

exports[`memo should memoize a function, sync errors 1`] = `
[
  {
    "cacheLength": 4,
    "errorValue": "[EMPTY]",
    "successValue": undefined,
    "type": "memo error",
  },
  {
    "cacheLength": 4,
    "errorValue": "[EMPTY]",
    "successValue": undefined,
    "type": "memo error",
  },
  {
    "cacheLength": 4,
    "errorValue": "4",
    "successValue": undefined,
    "type": "memo error",
  },
  {
    "cacheLength": 4,
    "errorValue": "4",
    "successValue": undefined,
    "type": "memo error",
  },
  {
    "cacheLength": 4,
    "errorValue": "5",
    "successValue": undefined,
    "type": "memo error",
  },
  {
    "cacheLength": 4,
    "errorValue": "5",
    "successValue": undefined,
    "type": "memo error",
  },
  {
    "cacheLength": 4,
    "errorValue": "6",
    "successValue": undefined,
    "type": "memo error",
  },
  {
    "cacheLength": 4,
    "errorValue": "6",
    "successValue": undefined,
    "type": "memo error",
  },
  {
    "cacheLength": 4,
    "errorValue": "4",
    "successValue": undefined,
    "type": "memo error",
  },
]
`;

exports[`memo should memoize a function, sync errors NOT cached 1`] = `
[
  {
    "cacheLength": 2,
    "errorValue": "7",
    "successValue": undefined,
    "type": "memo error",
  },
  {
    "cacheLength": 2,
    "errorValue": "7",
    "successValue": undefined,
    "type": "memo error",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "8",
    "type": "memo",
  },
  {
    "cacheLength": 4,
    "errorValue": undefined,
    "successValue": "8",
    "type": "memo",
  },
  {
    "cacheLength": 2,
    "errorValue": "9",
    "successValue": undefined,
    "type": "memo error",
  },
  {
    "cacheLength": 2,
    "errorValue": "9",
    "successValue": undefined,
    "type": "memo error",
  },
  {
    "cacheLength": 2,
    "errorValue": "7",
    "successValue": undefined,
    "type": "memo error",
  },
]
`;
